File: src/main/java/com/mindscapehq/raygun4android/RaygunMessageBuilder.java
package main.java.com.mindscapehq.raygun4android;

import android.content.Context;
import main.java.com.mindscapehq.raygun4android.messages.RaygunClientMessage;
import main.java.com.mindscapehq.raygun4android.messages.RaygunEnvironmentMessage;
import main.java.com.mindscapehq.raygun4android.messages.RaygunErrorMessage;
import main.java.com.mindscapehq.raygun4android.messages.RaygunMessage;

import java.util.AbstractList;
import java.util.Map;

public class RaygunMessageBuilder implements IRaygunMessageBuilder {

  private RaygunMessage _raygunMessage;

  public RaygunMessageBuilder()
  {
    _raygunMessage = new RaygunMessage();
  }

  public RaygunMessage Build()
  {
    return _raygunMessage;
  }

  @Override
  public IRaygunMessageBuilder SetMachineName(String machineName) {
    _raygunMessage.getDetails().setMachineName(machineName);
    return this;
  }

  @Override
  public IRaygunMessageBuilder SetExceptionDetails(Throwable throwable) {
    _raygunMessage.getDetails().setError(new RaygunErrorMessage(throwable));
    return this;
  }

  @Override
  public IRaygunMessageBuilder SetClientDetails() {
    _raygunMessage.getDetails().setClient(new RaygunClientMessage());
    return this;
  }

  @Override
  public IRaygunMessageBuilder SetEnvironmentDetails(Context context) {
    _raygunMessage.getDetails().setEnvironment(new RaygunEnvironmentMessage(context));
    return this;
  }

  @Override
  public IRaygunMessageBuilder SetVersion(String version) {
    _raygunMessage.getDetails().setVersion(version);
    return this;
  }

  @Override
  public IRaygunMessageBuilder SetTags(AbstractList tags) {
    _raygunMessage.getDetails().setTags(tags);
    return this;
  }

  @Override
  public IRaygunMessageBuilder SetUserCustomData(Map userCustomData) {
    _raygunMessage.getDetails().setUserCustomData(userCustomData);
    return this;
  }

  public static RaygunMessageBuilder New() {
    return new RaygunMessageBuilder();
  }


}

File: provider/src/main/java/com/raygun/raygun4android/RaygunMessageBuilder.kt
package com.raygun.raygun4android

import android.content.Context
import com.raygun.raygun4android.messages.crashreporting.RaygunBreadcrumbMessage
import com.raygun.raygun4android.messages.crashreporting.RaygunClientMessage
import com.raygun.raygun4android.messages.crashreporting.RaygunEnvironmentMessage
import com.raygun.raygun4android.messages.crashreporting.RaygunErrorMessage
import com.raygun.raygun4android.messages.crashreporting.RaygunMessage
import com.raygun.raygun4android.messages.shared.RaygunUserInfo

class RaygunMessageBuilder {
    private val raygunMessage = RaygunMessage()

    fun build(): RaygunMessage = raygunMessage

    fun setMachineName(machineName: String?): RaygunMessageBuilder {
        raygunMessage.details.machineName = machineName
        return this
    }

    fun setExceptionDetails(throwable: Throwable?): RaygunMessageBuilder {
        raygunMessage.details.error = RaygunErrorMessage(throwable)
        return this
    }

    fun setClientDetails(): RaygunMessageBuilder {
        raygunMessage.details.client = RaygunClientMessage()
        return this
    }

    suspend fun setEnvironmentDetails(context: Context): RaygunMessageBuilder {
        raygunMessage.details.environment = RaygunEnvironmentMessage(context)
        return this
    }

    fun setVersion(version: String?): RaygunMessageBuilder {
        raygunMessage.details.version = version
        return this
    }

    fun setTags(tags: List<*>?): RaygunMessageBuilder {
        raygunMessage.details.tags = tags
        return this
    }

    fun setCustomData(customData: Map<*, *>?): RaygunMessageBuilder {
        raygunMessage.details.customData = customData
        return this
    }

    fun setAppContext(identifier: String?): RaygunMessageBuilder {
        raygunMessage.details.setAppContext(identifier)
        return this
    }

    fun setUserInfo(raygunUserInfo: RaygunUserInfo): RaygunMessageBuilder {
        raygunMessage.details.user = raygunUserInfo
        return this
    }

    fun setNetworkInfo(context: Context): RaygunMessageBuilder {
        raygunMessage.details.setNetworkInfo(context)
        return this
    }

    fun setGroupingKey(groupingKey: String?): RaygunMessageBuilder {
        raygunMessage.details.groupingKey = groupingKey
        return this
    }

    fun setBreadcrumbs(breadcrumbs: List<RaygunBreadcrumbMessage>?): RaygunMessageBuilder {
        raygunMessage.details.breadcrumbs = breadcrumbs
        return this
    }
}
